### PR TITLE
IgnFindOGRE2: support for the ogre-next package on Ubuntu Jammy

### LINKED ISSUE #208: IgnOGRE2 support for ogre-next in Ubuntu Jammy
Ubuntu jammy provides a [libogre-next package ](https://packages.ubuntu.com/uk/jammy/libogre-next-dev) that ships a custom CMake helper that can potentially work mostly on its own (see the [patch](https://github.com/ignition-release/ign-rendering6-release/pull/8/files#diff-93975c3b98779e4d395782d26e01fdfe517a184b37ed202723eb0c57ebcb6cfd) used for ign-rendering code).

## Desired behavior
IgnOGRE2 to support the Ubuntu Jammy ogre-next installation.

## Alternatives considered
Until now, the alternative is to patch code to avoid IgnOGRE2 and use CMake diretly.

## Implementation suggestion
Try to support the Ubuntu jammy versions transparently with the other installations existing out there.

## Additional context
The root of the problem is the design of OGRE-Next to be an embedded piece of software inside other projects. This is hard to workaround and the origin of the IgnOGRE2 module.

### PR DESCRIPTION
**Pull request by j-rivero**
# 🎉 New feature

Closes #208

To review please remember to enable the GUI option to skip white-spaces as detailed in https://cloudfour.com/thinks/quick-tip-how-to-hide-whitespace-changes-in-git-diffs/

## Summary
Provides support to find OGRE2 now known as OGRE-Next where the code comes from the work done in Debian/Ubuntu (starting from Jammy) to package it with the name of OGRE-Next.

Workaround was in place for ign-rendering coded in https://github.com/ignitionrobotics/ign-rendering/pull/577 (although some libraries are missing for the correct linking). This PR implements the support to find that OGRE-Next installations on UNIX systems using the same call to FindIgnOGRE2 module.

~~The PR adds a new CMake input parameter to FindIgnOGRE2 module named `IGN_OGRE2_PROJECT_NAME` with default to OGRE2 that should keep the same behavior. The implementation idea is to keep using the same code for both flavors (OGRE2 and OGRE-Next) by creating new variables to deal with the difference~~

~~The output of the module are exactly the same variable names and components than before.~~

~~For a future implementation, we could probably apply some simple heuristics to auto-detect the the value of `IGN_OGRE2_PROJECT_NAME`.~~

The PR can autodetect which one of the package names is being used in the system OGRE2/OGRE-Next by iterating trying to find them one by one.

## Test it

You can use the [example available in this repository](https://github.com/ignitionrobotics/ign-cmake/tree/ign-cmake2/examples/find_ogre2/ogre-2.2) to check for output on the different Ubuntu distributions. ~~On jammy remember to use `-DIGN_OGRE2_PROJECT_NAME=OGRE-Next` when calling CMake.~~

For a full working ign-gazebo on Jammy I would recommend to use this branch with ign-rendering branch https://github.com/ignitionrobotics/ign-rendering/pull/605 (reverts previous workaround) using a colcon workspace inside a rocker run. ~~On jammy remember to use `--cmake-args -DIGN_OGRE2_PROJECT_NAME=OGRE-Next` when calling CMake.~~

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: cmake/FindIgnOGRE2.cmake [hunk 1]
```diff
@@ -22,6 +22,15 @@
...
 #     ign_find_package(IgnOGRE2)
 #
+# Variables used by this module, they can change the default behaviour and need
+# to be set before calling find_package:
+#
+#  IGN_OGRE2_PROJECT_NAME    Possible values: OGRE2 (default) or OGRE-Next
+#                            (Only on UNIX, not in use for Windows)
+#                            Specify the project name used in the packaging.
+#                            It will impact directly in the name of the
+#                            CMake/pkg-config modules beind used.
```
Comment (mjcarroll) [inline cmake/FindIgnOGRE2.cmake:32]:
```suggestion
#                            CMake/pkg-config modules being used.
```

**Review by mjcarroll — COMMENTED** [review]
Gave it a look over and looks good, still need to test it.

File: cmake/FindIgnOGRE2.cmake [hunk 2]
```diff
@@ -58,14 +68,34 @@ if (${IgnOGRE2_FIND_VERSION_MAJOR})
...
 message(STATUS " Version: 2.${IgnOGRE2_FIND_VERSION_MINOR}")
-set(OGRE2_INSTALL_PATH "OGRE-2.${IgnOGRE2_FIND_VERSION_MINOR}")
 
 macro(append_library VAR LIB)
   if(EXISTS "${LIB}")
     list(APPEND ${VAR} ${LIB})
   endif()
 endmacro()
 
+# Ubuntu Jammy version 2.2.5+dfsg3-0ubuntu2 is buggy in the pkg-config files
```
[the remark this answers is not in the document]
Comment (j-rivero) [reply cmake/FindIgnOGRE2.cmake:78]: I did not create it to centralized the patch in a single bug for Ubuntu using just ign-rendering https://bugs.launchpad.net/ubuntu/+source/ignition-rendering/+bug/1967888.

File: cmake/FindIgnOGRE2.cmake [hunk 3]
```diff
@@ -58,14 +68,34 @@ if (${IgnOGRE2_FIND_VERSION_MAJOR})
...
+  if (NOT EXISTS "${FILESYSTEM_PATH}")
+    string(REPLACE "/usr//usr" "/usr"
+      ${OUTPUT_VAR}
+      ${FILESYSTEM_PATH})
+  endif()
+endmacro()
+
+# Ubuntu Jammy version 2.2.5+dfsg3-0ubuntu2 is buggy in the pkg-config files
+# using a non existing path /usr/lib/${arch}/OGRE/OGRE-Next insted of the path
+# /usr/lib/${arch}/OGRE-Next which is the right one
```
[the remark this answers is not in the document]
Comment (j-rivero) [reply cmake/FindIgnOGRE2.cmake:90]: same than above, no.

**Comment by chapulina** [comment]
> For a full working ign-gazebo on Jammy I would recommend to use this branch with ign-rendering branch jrivero/6_revert_ogrenext_workaround

Are you planning to open an `ign-rendering` PR for this? Does this PR work with `ign-rendering`'s current `main`?

**Comment by j-rivero** [comment]
While working on getting out-of-the-box support for ign-rendering I ended up implementing the autodetection of OGRE2/OGRE-Next in this module [4e7f535](https://github.com/ignitionrobotics/ign-cmake/pull/224/commits/4e7f535a17065ca35f227324a7a651443dc032cd). First it tries to locate the software under OGRE2 name and if it fails under OGRE-Next name. This in combination with https://github.com/ignitionrobotics/ign-rendering/pull/605 (which run the detection twice if failed to locate PlanarReflections) produces the following:

Focal
```bash
-- Looking for OGRE using the name: OGRE2
--   + component HlmsPbs: found
--   + component HlmsUnlit: found
--   + component Overlay: found
--   ! component PlanarReflections: not found!
--   ! OGRE2 not found
-- Looking for OGRE using the name: OGRE-Next
--   ! OGRE-Next not found                              
-- Looking for IgnOGRE2 - not found

-- IgnOGRE2 was not found with PlanarReflections. Try looking without this component:
-- Looking for OGRE using the name: OGRE2
--   + component HlmsPbs: found
--   + component HlmsUnlit: found
--   + component Overlay: found
-- Looking for IgnOGRE2 - found
```

Jammy
```bash
-- Looking for OGRE using the name: OGRE2
--   ! OGRE2 not found                                  
-- Looking for OGRE using the name: OGRE-Next
--   + component HlmsPbs: found
--   + component HlmsUnlit: found
--   + component Overlay: found
--   + component PlanarReflections: found
-- Looking for IgnOGRE2 - found
```

**Review by mjcarroll — APPROVED** [review]

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-releases-2022-04-27-fortress-citadel/1389/1

**Pull request merged**